python/tvm/relax/op/_op_gradient.py

```diff
@@ -0,0 +1,488 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=unused-argument
+"""Gradient definitions for Relax operators"""
+from typing import List
+from tvm import relax
+from tvm.arith import Analyzer
+from tvm.relax.expr import Call, Var, Expr, ShapeExpr
+from tvm._ffi.base import TVMError
+
+from ..block_builder import BlockBuilder
+from ...tir import PrimExpr
+from .base import register_gradient
+
+from .unary import (
+    log,
+    negative,
+    exp,
+)
+from .binary import (
+    subtract,
+    multiply,
+    divide,
+    less,
+)
+from .statistical import sum as _sum
+from .statistical import mean
+from .create import (
+    zeros,
+    ones,
+)
+from .search import where
+from .linear_algebra import matmul
+from .manipulate import (
+    collapse_sum_to,
+    broadcast_to,
+    permute_dims,
+    expand_dims,
+    concat,
+    split,
+)
+
+
+def _get_shape(expr: Expr) -> ShapeExpr:
+    try:
+        shape = expr.struct_info.shape
+    except Exception as error:
+        raise TVMError(
+            f"Get the shape of {expr} failed. Please normalize it first and ensure it is a Tensor."
+        ) from error
+    return shape
+
+
+def _fit_shape(expr: Expr, expr_shape: ShapeExpr, target: Expr) -> Expr:
+    target_shape = _get_shape(target)
+
+    def _check_shape_equal():
+        expr_sinfo = expr_shape.struct_info
+        target_sinfo = target_shape.struct_info
+        if len(expr_sinfo.values) != len(target_sinfo.values):
+            return False
+        analyzer = Analyzer()
+        for i, field in enumerate(expr_sinfo.values):
+            if not analyzer.can_prove_equal(field, target_sinfo.values[i]):
+                return False
+        return True
```

Comment:
We have similar shape analysis function
https://github.com/mlc-ai/relax/blob/760dbe8a71361654fa64fa5f2d8a31f777f7d1ac/include/tvm/relax/analysis.h#L43-L70

But it isn’t exposed to FFI and Python side, and it also assumes the input Expr are ShapeExpr. You can add some new function to .h/.cc/.py accordingly for your case.

Reply:
Let me expose it to FFI and call it

Reply:
There are some problems about exposing it to python. Let's just keep it for now